core/src/apps/zcash/orchard/crypto/note_encryption.py

```diff
@@ -0,0 +1,120 @@
+import gc
+from micropython import const
+from typing import TYPE_CHECKING
+
+from trezor.crypto import chacha20poly1305
+from trezor.crypto.hashlib import blake2b
+from trezor.crypto.pallas import Point
+from trezor.utils import empty_bytearray, ensure
+
+from apps.common.writers import write_bytes_fixed
+
+from .note import Note
+
+if TYPE_CHECKING:
+    from typing import Iterable
+    from ..random import ActionShieldingRng
+
+BLOCK_SIZE = const(64)
+ENC_CIPHERTEXT_SIZE = const(580)
+OUT_CIPHERTEXT_SIZE = const(80)
+
+
+# https://zips.z.cash/protocol/nu5.pdf#concreteorchardkdf
+def kdf_orchard(shared_secret: Point, ephemeral_key: bytes) -> bytes:
+    digest = blake2b(outlen=32, personal=b"Zcash_OrchardKDF")
+    digest.update(shared_secret.to_bytes())
+    digest.update(ephemeral_key)
+    return digest.digest()
+
+
+# https://zips.z.cash/protocol/nu5.pdf#concreteprfs
+def prf_ock_orchard(
+    ovk: bytes,
+    cv: bytes,
+    cmx: bytes,
+    ephemeral_key: bytes,
+) -> bytes:
+    digest = blake2b(outlen=32, personal=b"Zcash_Orchardock")
+    digest.update(ovk)
+    digest.update(cv)
+    digest.update(cmx)
+    digest.update(ephemeral_key)
+    return digest.digest()
+
+
+def chunks(size: int, length: int) -> Iterable[tuple[int, int]]:
+    offset = 0
+    while offset + size <= length:
+        yield offset, offset + size
+        offset += size
+    if offset < length:
+        yield offset, length
+    return
+
+
+# https://zips.z.cash/protocol/nu5.pdf#concretesym
+def sym_encrypt(key: bytes, buffer: bytearray) -> None:
+    nonce = 12 * b"\x00"
+    cipher = chacha20poly1305(key, nonce)
+    for i, j in chunks(BLOCK_SIZE, len(buffer)):
+        buffer[i:j] = cipher.encrypt(buffer[i:j])
+    buffer.extend(cipher.finish())  # append tag
+
+
+class TransmittedNoteCiphertext:
+    def __init__(
+        self,
+        epk_bytes: bytes,
+        enc_ciphertext: bytes,
+        out_ciphertext: bytes,
+    ):
+        self.epk_bytes = epk_bytes
+        self.enc_ciphertext = enc_ciphertext
+        self.out_ciphertext = out_ciphertext
+
+
+# https://zips.z.cash/protocol/nu5.pdf#saplingandorchardencrypt
+def encrypt_note(
+    note: Note,
+    memo: str | bytes | None,
+    cv_new: Point,
+    cm_new: Point,
+    ovk: bytes | None,
+    rng: ActionShieldingRng,
+) -> TransmittedNoteCiphertext:
+    np = empty_bytearray(ENC_CIPHERTEXT_SIZE)
```

Comment:
a better choice is to make `_NP` a preallocated global and only clear it here, dtto for `op`/`_OP`

Reply:
0ca4bfce7ab96a4ba2bd835ffc45225cdd3cc5a2

Reply:
check that I'm doing it right please

Reply:
that looks correct. I don't think you need the `global` declaration, because you never reassign the `_NP`/`_OP` symbol

Reply:
please add a comment saying that these are static buffers for `encrypt_note`

Reply:
> I don't think you need the global declaration, because you never reassign the _NP/_OP symbol
right . then a global declaration of `_SERIALIZED_TX_BUFFER` in `Bitcoin.__init__` is not necessary neither

Reply:
fab42f84b9c2406ed066726bf5a14a93f9d1f471